rir/src/compiler/opt/types.cpp

```diff
@@ -40,33 +54,25 @@ void TypeInference::apply(RirCompiler&, ClosureVersion* function,
                     i->eachArg([&](Value* v) {
                         if (i->mayHaveEnv() && v == i->env())
                             return;
-                        if (auto arg = Instruction::Cast(v)) {
-                            if (types.count(arg))
-                                infered = infered | types.at(arg);
-                            else
-                                done = false;
-                        } else {
-                            infered = infered | v->type;
-                        }
+                        infered = infered | getType(v);
                     });
                     if (i->tag == Tag::Div && infered.isA(RType::integer)) {
                         infered = infered | RType::real;
                     }
                     break;
                 }
-                case Tag::Assume: {
-                    auto assumption = Assume::Cast(i);
-                    if (!assumption->assumeTrue)
-                        if (auto isO = IsObject::Cast(assumption->condition()))
-                            if (auto val =
-                                    Instruction::Cast(isO->arg<0>().val())) {
-                                if (types.count(val))
-                                    infered = types.at(val).notObject();
-                                else
-                                    infered = val->type.notObject();
-                            }
-                    break;
+                case Tag::CallSafeBuiltin: {
```

Comment:
Why was the `Assume` case removed?

Reply:
right, we used to just do the following:
```
t = IsObject(o)
Assume(t)
...
```
but with the type speculation branch, I actually changed this to
```
t = IsObject(o)
Assume(t)
t1 = CastType(t, t->type.notObject())
...
```
and then replace t with t1. so this case is just not needed anymore, since the assumption is already encoded in the castType.